pkg/azuredx/cache.go

```diff
@@ -0,0 +1,111 @@
+package azuredx
+
+import (
+	"fmt"
+	"regexp"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+)
+
+const cacheMinAge = time.Duration(10 * time.Second)
+const cacheMaxAge = time.Hour
+
+// CacheSettings is use to enable caching on a per query basis
+type CacheSettings struct {
+	CacheMaxAge string
+	TimeRange   *backend.TimeRange
+}
+
+// NewCacheSettings is used to detect what cache settings is applicable for the current query.
+func NewCacheSettings(c *Client, q *backend.DataQuery, qm *QueryModel) *CacheSettings {
+	if !c.DynamicCaching {
+		return &CacheSettings{
+			CacheMaxAge: c.CacheMaxAge,
+			TimeRange:   &q.TimeRange,
+		}
+	}
+
+	resolution, err := detectBinSize(qm.Query, q.Interval)
+	if err != nil {
+		resolution = detectResolution(&q.TimeRange)
+	}
+
+	maxAge := adjustToMaxMinLimits(resolution)
+
+	return &CacheSettings{
+		CacheMaxAge: fmt.Sprintf("%vs", maxAge.Seconds()),
+		TimeRange:   expandTimeRange(&q.TimeRange, maxAge),
+	}
+}
+
+func expandTimeRange(tr *backend.TimeRange, d time.Duration) *backend.TimeRange {
+	return &backend.TimeRange{
+		From: expandFrom(tr.From, d),
+		To:   expandTo(tr.To, d),
+	}
+}
+
+func expandFrom(from time.Time, d time.Duration) time.Time {
+	expanded := from.Round(d)
+
+	if expanded.After(from) {
+		return expanded.Add(d * -1)
+	}
+	return expanded
+}
+
+func expandTo(to time.Time, d time.Duration) time.Time {
+	expanded := to.Round(d)
+
+	if expanded.Before(to) {
+		return expanded.Add(d)
+	}
+	return expanded
+}
+
+func adjustToMaxMinLimits(resolution time.Duration) time.Duration {
+	if resolution > cacheMaxAge {
+		return cacheMaxAge
+	}
+	if resolution < cacheMinAge {
+		return cacheMinAge
+	}
+	return resolution
+}
+
+func detectResolution(tr *backend.TimeRange) time.Duration {
+	diff := tr.To.Sub(tr.From)
+
+	if diff > time.Hour*24*7*4 {
+		return cacheMaxAge
+	}
+
+	if diff > time.Hour {
+		return time.Minute
+	}
+
+	return cacheMinAge
+}
+
+// binSizeRE finds the bin size of a query by looking at the summarize statement
+// e.g. summarize avg(Column), count(Column) by bin(TimeColumn, 1d)
+var binSizeRE = regexp.MustCompile(`summarize [\w\$\(\)\.,\s]+ by bin\([\w\$\(\)\.]+, ` +
```

Comment:
Worry about change query behavior based on a regex. But if we have to do it, I think we should just detect the bin.